packages/flutter/lib/src/widgets/animated_list.dart

```diff
@@ -0,0 +1,218 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/animation.dart';
+import 'package:flutter/foundation.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'scroll_view.dart';
+import 'ticker_provider.dart';
+
+typedef Widget AnimatedListItemBuilder(BuildContext context, int index, Animation<double> animation);
+typedef Widget AnimatedListRemovedItemBuilder(BuildContext context, Animation<double> animation);
+typedef void AnimatedListInitialItemsBuilder(AnimatedListState list);
+
+// The default insert/remove animation duration.
+const Duration _kDuration = const Duration(milliseconds: 3000);
+
+class _ActiveItem implements Comparable<_ActiveItem> {
+  _ActiveItem.incoming(this.controller, this.itemIndex) : removedItemBuilder = null;
+  _ActiveItem.outgoing(this.controller, this.itemIndex, this.removedItemBuilder);
+
+  final AnimationController controller;
+  final AnimatedListRemovedItemBuilder removedItemBuilder;
+  int itemIndex;
+
+  @override
+  int compareTo(_ActiveItem other) => itemIndex - other.itemIndex;
+}
+
+/// A scrolling container that animates items when they are inserted or removed.
+///
+/// This widget's [AnimatedListState] can be used to insert or remove items. To
+/// refer to the [AnimatedListState] either provide a [GlobalKey] or use
+/// the static [of] method from a item's input callback.
+class AnimatedList extends StatefulWidget {
+  AnimatedList({ Key key,  this.itemBuilder, this.initialItemCount: 0 }) : super(key: key) {
+    assert(itemBuilder != null);
+    assert(initialItemCount != null && initialItemCount >= 0);
+  }
+
+  final AnimatedListItemBuilder itemBuilder;
+  final int initialItemCount;
+
+  // TBD: explain that this is typically used by List item handlers that want to refer to the list.
+  static AnimatedListState of(BuildContext context, { bool nullOk: false }) {
+    assert(nullOk != null);
+    assert(context != null);
+    final AnimatedListState result = context.ancestorStateOfType(const TypeMatcher<AnimatedListState>());
+    if (nullOk || result != null)
+      return result;
+    throw new FlutterError(
+      'AnimatedList.of() called with a context that does not contain a AnimatedList.\n'
+      'No AnimatedList ancestor could be found starting from the context that was passed to AnimatedList.of(). '
+      'This can happen when the context provided is from the same StatefulWidget that '
+      'built the AnimatedList. Please see the AnimatedList documentation for examples '
+      'of how to refer to an AnimatedListState object: '
+      '  https://docs.flutter.io/flutter/widgets/AnimatedState-class.html\n'
```

Comment:
we probably want to flesh out these docs before we point to them. at least have some sample code or something. :-)